frontend/src/metabase-lib/parameters/utils/parameter-source.ts

```diff
@@ -1,27 +1,49 @@
-import { ValuesSourceConfig, ValuesSourceType } from "metabase-types/api";
+import {
+  Dataset,
+  ValuesSourceConfig,
+  ValuesSourceType,
+} from "metabase-types/api";
 
 export const isValidSourceConfig = (
   sourceType: ValuesSourceType,
-  sourceConfig: ValuesSourceConfig,
+  { card_id, value_field, values }: ValuesSourceConfig,
 ) => {
   switch (sourceType) {
     case "card":
-      return sourceConfig.card_id != null && sourceConfig.value_field != null;
+      return card_id != null && value_field != null;
     case "static-list":
-      return sourceConfig.values != null && sourceConfig.values.length > 0;
+      return values != null && values.length > 0;
     default:
       return true;
   }
 };
 
-export const getDefaultSourceConfig = (
+export const getSourceConfigForType = (
   sourceType: ValuesSourceType,
-  sourceValues?: string[],
+  { card_id, value_field, values }: ValuesSourceConfig,
 ): ValuesSourceConfig => {
   switch (sourceType) {
+    case "card":
+      return { card_id, value_field };
     case "static-list":
-      return { values: sourceValues };
+      return { values };
     default:
       return {};
   }
 };
+
+const getUniqueNonNullValues = (values: unknown[]) => {
+  return Array.from(new Set(values))
+    .filter(value => value != null)
+    .map(value => String(value));
+};
+
+export const getFieldSourceValues = (fieldsValues: unknown[][][]) => {
```

Comment:
could the type be more descriptive? seeing nested arrays 3 times looks really confusing to me as a person who's not familiar with field values structure.

Reply:
Exported `FieldValue` structure and used it here